### PR TITLE
Death range component

### PR DESCRIPTION
**Pull request by tibaudlemaout**
new death range component to kill projectile entities that move past a min or max x position value
changed the entity assignation of IDs to keep track of deleted one to be reused
SparseArray is now a vector instead of an unordered_map

### CODE REVIEW

**Review by Tugduoff — DISMISSED** [review]

File: plugins/systems/input/Input.hpp [hunk 1]
```diff
@@ -54,6 +55,8 @@ namespace Systems {
                 auto &reg = engine.getRegistry();
                 ECS::Entity projectile = reg.createEntity();
 
+                std::cerr << "New entity for projectile" << projectile << std::endl;
```
Comment (Tholarr) [inline plugins/systems/input/Input.hpp:58]: debug print or usefull print ?

File: plugins/systems/move/Move.cpp [hunk 2]
```diff
@@ -24,26 +25,50 @@ void Systems::MoveSystem::run(Engine::GameEngine &engine)
...
+                std::cerr << "Error: " << e.what() << std::endl;
                 continue;
-            pos->x += vel->x;
-            pos->y += vel->y;
-            engine.updateComponent(i, pos->getId(), pos->serialize());
-            if (vel->diminishingFactor == 0)
+            }
+        }
+        // std::cerr << "DeathRange size: " << deathRangeComponent.size() << std::endl;
+        // std::cerr << "Position size: " << posComponents.size() << std::endl;
```
Comment (Tholarr) [inline plugins/systems/move/Move.cpp:53]: comments left on purpose ?

File: plugins/systems/move/Move.cpp [hunk 3]
```diff
@@ -24,26 +25,50 @@ void Systems::MoveSystem::run(Engine::GameEngine &engine)
...
+        for (i = 0; i < posComponents.size() && i < deathRangeComponent.size(); i++) {
+            std::cerr << "Checking death range for entity " << i << std::endl;
```
Comment (Tholarr) [inline plugins/systems/move/Move.cpp:55]: debug print ?

File: plugins/systems/move/Move.cpp [hunk 4]
```diff
@@ -24,26 +25,50 @@ void Systems::MoveSystem::run(Engine::GameEngine &engine)
...
+            try {
+                auto &drange = deathRangeComponent[i];
+                auto &pos = posComponents[i];
+
+                if (!pos || !drange)
+                    continue;
+                if (pos->x > drange->maximum || pos->x < drange->minimum) {
+                    reg.killEntity((ECS::Entity)i);
+                    std::cerr << "Entity " << i << " has been killed due to death range" << std::endl;
```
Comment (Tholarr) [inline plugins/systems/move/Move.cpp:64]: debug print ?

File: plugins/systems/move/Move.cpp [hunk 5]
```diff
@@ -24,26 +25,50 @@ void Systems::MoveSystem::run(Engine::GameEngine &engine)
...
+                }
+            } catch (std::exception &e) {
+                std::cerr << "Error: " << e.what() << std::endl;
```
Comment (Tholarr) [inline plugins/systems/move/Move.cpp:67]: debug print ?

File: plugins/systems/sfml/Display.cpp [hunk 6]
```diff
@@ -65,6 +65,7 @@ void Systems::Display::run(Engine::GameEngine &engine, sf::RenderWindow &window)
     window.clear();
 
     try {
+        std::cerr << "DisplaySystem running" << std::endl;
```
Comment (Tholarr) [inline plugins/systems/sfml/Display.cpp:68]: debug print ?

File: plugins/systems/sfml/Display.cpp [hunk 7]
```diff
@@ -77,25 +78,34 @@ void Systems::Display::run(Engine::GameEngine &engine, sf::RenderWindow &window)
...
                 try {
                     auto &sprite = spriteComponents[i];
                     (void)sprite;
                 } catch (std::exception &) {
                     std::cerr << "Error: Sprite component not found for this entity, creating it..." << std::endl;
                     std::unique_ptr<Components::SpriteComponent> spriteComp = std::make_unique<Components::SpriteComponent>();
                     reg.componentManager().addComponent<Components::SpriteComponent>((ECS::Entity)i, std::move(spriteComp));
+                    break;
                 }
+                std::cerr << "Entity " << i << " has a sprite component" << std::endl;
```
Comment (Tholarr) [inline plugins/systems/sfml/Display.cpp:105]: debug print ?

File: plugins/systems/input/Input.hpp (hunk 1 above)
Comment (tibaudlemaout) [reply plugins/systems/input/Input.hpp:58]: It is useful to keep track of the IDs used for creating projectile, to see how loaded the vectors can get

File: plugins/systems/sfml/Display.cpp [hunk 8]
```diff
@@ -77,25 +78,34 @@ void Systems::Display::run(Engine::GameEngine &engine, sf::RenderWindow &window)
...
                 auto &sprite = spriteComponents[i];
+                std::cerr << "yu" << std::endl;
```
Comment (Tholarr) [inline plugins/systems/sfml/Display.cpp:107]: yu ? x)
debug print ?

File: plugins/systems/sfml/Display.cpp [hunk 9]
```diff
@@ -125,6 +135,7 @@ void Systems::Display::run(Engine::GameEngine &engine, sf::RenderWindow &window)
                 window.draw(sprite->sprite);
             }
         }
+        std::cerr << "End of display system" << std::endl;
```
Comment (Tholarr) [inline plugins/systems/sfml/Display.cpp:138]: debug print ?

File: plugins/systems/move/Move.cpp (hunk 4 above)
Comment (tibaudlemaout) [reply plugins/systems/move/Move.cpp:64]: This one is useful to make sure the right Entity has been killed, as we had issues with swapping Entity indexes

**Review by Tugduoff — APPROVED** [review]

**Review by Tholarr — APPROVED** [review]
seems nice :saluting_face:

**Pull request merged**